Documentation/network/lb-ipam.rst

```diff
@@ -0,0 +1,382 @@
+
+Lets say we add a simple service.
+
+.. code-block:: yaml
+
+    apiVersion: v1
+    kind: Service
+    metadata:
+      name: service-red
+      namespace: example
+      labels:
+        color: red
+    spec:
+      type: LoadBalancer
+      ports:
+      - port: 1234
+
+This service will appear like so.
+
+.. code-block:: shell-session
+
+    $ kubectl -n example get svc
+    NAME          TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
+    service-red   LoadBalancer   10.96.192.212   <pending>     1234:30628/TCP   24s
+
+The ExternalIP field has a value of ``<pending>`` which means no LB IPs have been assigned.
+When LB IPAM is unable to allocate or assign IPs for the service, it will update the service
+conditions in the status.
+
+The service conditions can be checked like so:
+
+.. code-block:: shell-session
+
+    kubectl -n example get svc/service-red -o jsonpath='{.status.conditions}' | jq
+    [
+      {
+        "lastTransitionTime": "2022-10-06T13:40:48Z",
+        "message": "There are no enabled CiliumLoadBalancerIPPools that match this service",
+        "reason": "no_pool",
+        "status": "False",
+        "type": "io.cilium/lb-ipam-request-satisfied"
+      }
+    ]
+
+After updating the service labels to match our ``blue-pool`` from before we see:
+
+.. code-block:: shell-session
+
+    $ kubectl -n example get svc
+    NAME          TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
+    service-red   LoadBalancer   10.96.192.212   20.0.10.163   1234:30628/TCP   12m
+
+    $ kubectl -n example get svc/service-red -o jsonpath='{.status.conditions}' | jq
+    [
+      {
+        "lastTransitionTime": "2022-10-06T13:40:48Z",
+        "message": "There are no enabled CiliumLoadBalancerIPPools that match this service",
+        "reason": "no_pool",
+        "status": "False",
+        "type": "io.cilium/lb-ipam-request-satisfied"
+      },
+      {
+        "lastTransitionTime": "2022-10-06T13:52:55Z",
+        "message": "",
+        "reason": "satisfied",
+        "status": "True",
+        "type": "io.cilium/lb-ipam-request-satisfied"
+      }
+    ]
+
+IPv4 / IPv6 families + policy
+-----------------------------
+
+LB IPAM supports IPv4 and/or IPv6 in SingleStack or `DualStack <https://kubernetes.io/docs/concepts/services-networking/dual-stack/>`__ mode. 
+Services can use the ``.spec.ipFamilyPolicy`` and ``.spec.ipFamilies`` fields to change
+the requested IPs.
+
+If ``.spec.ipFamilyPolicy`` isn't specified, ``SingleStack`` mode is assumed. 
+If both IPv4 and IPv6 are enabled in ``SingleStack`` mode, an IPv4 address is allocated.
+
+If ``.spec.ipFamilyPolicy`` is set to ``PreferDualStack``, LB IPAM will attempt to allocate 
+both an IPv4 and IPv6 address if both are enabled on the cluster. If only IPv4 or only IPv6 is
+enabled on the cluster, the service is still considered "satisfied".
+
+If ``.spec.ipFamilyPolicy`` is set to ``RequireDualStack`` LB IPAM will attempt to allocate
```

Comment:
In the dual-stack case, could it allocate addresses from disparate pools? Or should we require that both allocations for a a given service come from the same pool?

I think it's probably OK to use disparate pools, but I want to consider all the implications.

Reply:
We currently allow allocations from disparate pools. We have no requirement or use case for explicitly allocating from the same pool, perhaps one day.

There aren't much implications that I know of. If one of the pools or ranges is deleted, LB-IPAM will attempt to allocate a new IP. There is a chance that fails, in which case a service might only have a IPv4 or IPv6 address. In any case, this should be reflected in the services conditions and thus should be noticeable and can be monitored.